### PR TITLE
pinning API

### PR DESCRIPTION
**Pull request by diasdavid**
> **WIP**

- API calls
  - [x] ipfs.pin.ls
  - [x] ipfs.pin.add
  - [x] ipfs.pin.rm
- [ ] ~~Works in js-ipfs~~ milestone 6
- [x] Works in js-ipfs-api

### CODE REVIEW

File: API/pinning-api/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+pinning API
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: Pinning

File: API/pinning-api/README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+##### `Go` **WIP**
+
+##### `JavaScript` - ipfs.pin.add(hash, [options, callback])
+
+Where:
+
+- `hash` is an IPFS multihash.
+- `options` is an object that can contain the following keys:
+  - `recursive` - Recursevely pin the object linked.
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: recursively

Comment (RichardLitt) [inline API/pinning-api/README.md]: Default? Type bool?

File: API/pinning-api/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+`callback` must follow `function (err) {}` signature, where `err` is an error if the operation was not successful.
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: No res?

File: API/pinning-api/README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+##### `Go` **WIP**
+
+##### `JavaScript` - ipfs.pin.ls([hash, options, callback])
+
+Where:
+
+- `hash` is an IPFS multihash.
+- `options` is an object that can contain the following keys:
+  - `type` - Return also the type of pin (direct, indirect or recursive)
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: What about `all`? Should be the default. 

Put these in backticks, they are the strings to pass, not vars.

File: API/pinning-api/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+`callback` must follow `function (err, pinset) {}` signature, where `err` is an error if the operation was not successful. `pinset` is an array of objects with keys `hash` and `type`.
+
+If no `callback` is passed, a promise is returned.
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: After pinset, or before?

File: API/pinning-api/README.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+Example:
+
+```JavaScript
+ipfs.pin.ls(function (err, pinset) {})
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: should what the callback would normally do?

File: API/pinning-api/README.md (hunk 4 above)
Comment (RichardLitt) [inline API/pinning-api/README.md]: What about `quiet`? 

> Write just hashes of objects. Default: false.

File: API/pinning-api/README.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+```
+
+
+#### `rm`
+
+> Remove an hash from the pinset
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: A hash

Comment (RichardLitt) [inline API/pinning-api/README.md]: Should be: 
> Removes the pinned object from local storage.

Where is the term pinset coming from?

File: API/pinning-api/README.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+##### `Go` **WIP**
+
+##### `JavaScript` - ipfs.pin.rm(hash, [callback])
+
+Where `hash` is a multihash.
+
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: What about recursive? Should be a default true bool.

File: API/pinning-api/README.md (hunk 4 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: `quiet` is a cli option.

File: API/pinning-api/README.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+`callback` must follow `function (err) {}` signature, where `err` is an error if the operation was not successful.
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: Why no `pinset`? Why no promise?

File: package.json [hunk 10]
```diff
@@ -47,4 +47,4 @@
     "greenkeeperio-bot <support@greenkeeper.io>",
     "nginnever <ginneversource@gmail.com>"
   ]
-}
\ No newline at end of file
+}
```
Comment (RichardLitt) [inline package.json]: Unrelated. Pull out, put in own PR.

File: API/pinning-api/README.md (hunk 5 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: The promise is return, but resolved at the same time a callback would be.

File: API/pinning-api/README.md (hunk 6 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: Not sure if I understand

File: src/pin.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+const testfile = fs.readFileSync(path.join(__dirname, './data/testfile.txt'))
+
+module.exports = (common) => {
+  describe('.pin', () => {
+    let ipfs
+
+    before(function (done) {
+      // CI is slow
+      this.timeout(20 * 1000)
```
Comment (RichardLitt) [inline src/pin.js]: Factor this out, make it a global thing for all tests?

File: src/pin.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+      common.setup((err, factory) => {
```
Comment (RichardLitt) [inline src/pin.js]: Where does `common` come from? Curious.

File: API/pinning-api/README.md (hunk 7 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: pinset is the list of hashes that are pinned. Only ipfs pin rm and then ipfs repo gc actually removes the object, until gc is performed, the objects will be kept.

Comment (diasdavid) [inline API/pinning-api/README.md]: We've been using it for a while, it is part of the 'pinning' lingo

File: src/pin.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    })
+
+    it('add file for testing', (done) => {
+      const expectedMultihash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+      ipfs.files.add(testfile, (err, res) => {
+        expect(err).to.not.exist
+
+        expect(res).to.have.length(1)
+        expect(res[0].hash).to.equal(expectedMultihash)
```
Comment (RichardLitt) [inline src/pin.js]: Shouldn't this be a 'pinned <hash>.' or something? Do you just return the hash in js? Isn't that not feature parity?

File: src/pin.js [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        expect(res[0].path).to.equal(expectedMultihash)
+        done()
+      })
+    })
+
+    describe('callback API', () => {
+      it('.remove (first, because adding is recursive)', (done) => {
```
Comment (RichardLitt) [inline src/pin.js]: No. Remove should also be recursive.

File: API/pinning-api/README.md (hunk 7 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: added a note on the `add` to explain what pinset is :)

File: src/pin.js [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
```
Comment (RichardLitt) [inline src/pin.js]: Make this const declared above, not sure why you declare it so many times.

File: src/pin.js [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+        ipfs.pin.rm(hash, { recursive: true }, (err, res) => {
```
Comment (RichardLitt) [inline src/pin.js]: This isn't an option in the docs. Add it there?

File: src/pin.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+          expect(err).to.not.exist
+          expect(res).to.exist
+          ipfs.pin.ls({ type: 'direct' }, (err, res) => {
```
Comment (RichardLitt) [inline src/pin.js]: Try with all different types?

File: src/pin.js [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        return ipfs.pin.add(hash, { recursive: false })
+          .then((res) => {
+            expect(res.Pins[0]).to.be.equal(hash)
+          })
+      })
+
+      it('.ls', () => {
+        return ipfs.pin.ls()
+          .then((res) => {
+            expect(res).to.exist
```
Comment (RichardLitt) [inline src/pin.js]: probably not tight enough.

File: API/pinning-api/README.md (hunk 9 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: Promise is return if no callback is passed (says on line below)

File: src/pin.js (hunk 11 above)
Comment (diasdavid) [inline src/pin.js]: It is just the step of creating the node. If put globally, then we always have to wait 20 seconds for tests to timeout.

File: API/pinning-api/README.md (hunk 7 above)
Comment (RichardLitt) [inline API/pinning-api/README.md]: Cool. Yeah, less "What is the new term" and more "Where is it defined". Add it to the glossary? :D

File: src/pin.js (hunk 12 above)
Comment (diasdavid) [inline src/pin.js]: It is the 'interface test suite pattern', see: https://github.com/ipfs/interface-ipfs-core#usage

File: API/pinning-api/README.md (hunk 9 above)
Comment (RichardLitt) [inline API/pinning-api/README.md]: Why not `function (err, Promise) {}` then?

File: src/pin.js (hunk 13 above)
Comment (diasdavid) [inline src/pin.js]: Since we create an ephemeral node for each batch o test, we have to populate that node.

File: API/pinning-api/README.md (hunk 6 above)
Comment (RichardLitt) [inline API/pinning-api/README.md]: How about:

```js
ipfs.pin.ls(function (err, pinset) {
  if (err) return err
  return pinset 
})
```

That's the default callback, no?

File: src/pin.js (hunk 14 above)
Comment (diasdavid) [inline src/pin.js]: What I meant here was " since ipfs.files.add automatically pins the files, in order to properly test the remaining pin tests, we have to unpin first)"

File: src/pin.js (hunk 11 above)
Comment (RichardLitt) [inline src/pin.js]: Why not say that, instead of "CI is slow?"

File: API/pinning-api/README.md [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+- `hash` is an IPFS multihash.
+- `options` is an object that can contain the following keys
+  - `recursive` - Recursively pin the object linked. Type: bool. Default: 'false'
+
+`callback` must follow `function (err, res) {}` signature, where `err` is an error if the operation was not successful. `res` is an array of objects that represent the files that were pinned. Example:
+
+```JavaScript
+{
+  hash: 'QmHash',
+  path: 'some-path
```
Comment (RichardLitt) [inline API/pinning-api/README.md]: Missing '.

File: API/pinning-api/README.md (hunk 6 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: Ah, populate the example, ok :) you got it!

File: API/pinning-api/README.md (hunk 9 above)
Comment (diasdavid) [inline API/pinning-api/README.md]: Returning a Promise inside a callback would be an anti-pattern.

File: src/pin.js (hunk 11 above)
Comment (diasdavid) [inline src/pin.js]: good idea!

File: src/pin.js (hunk 15 above)
Comment (diasdavid) [inline src/pin.js]: I kind of like of having tests that are really self descriptive, it makes it easier to read and to change.

File: src/pin.js (hunk 16 above)
Comment (diasdavid) [inline src/pin.js]: thank you for catching that :)

**Comment by diasdavid** [comment]
Thank you @RichardLitt :D Applied almost all your CR, will add a couple more tests to cover remaining cases :)

File: API/pinning-api/README.md (hunk 9 above)
Comment (RichardLitt) [inline API/pinning-api/README.md]: Fair enough. Sometimes my mind is not all here.

File: src/pin.js (hunk 13 above)
Comment (RichardLitt) [inline src/pin.js]: What does the `res` look like, though?

Comment (diasdavid) [inline src/pin.js]: It is an array of objects, as defined in: https://github.com/ipfs/interface-ipfs-core/tree/master/API/files#javascript---ipfsfilesadddata-callback

Comment (RichardLitt) [inline src/pin.js]: Can you test the arr using deepEquals?

File: API/pinning-api/README.md [hunk 20]
```diff
@@ -0,0 +1,91 @@
+Pin API
```
Comment (dignifiedquire) [inline API/pinning-api/README.md]: The folder is called `pinning-api`, so this probably should be called `Pinning API`

File: API/pinning-api/README.md [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+If no `callback` is passed, a promise is returned.
+
+Example:
+
+```JavaScript
+ipfs.pin.ls(function (err, pinset) {
+  if (err) {
+    throw err
+  }
+  // console.log(pinset)
```
Comment (dignifiedquire) [inline API/pinning-api/README.md]: Why is this commented out?

File: src/pin.js [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+          done()
+        })
+      })
+    })
+
+    after((done) => {
+      common.teardown(done)
+    })
+
+    it('add file for testing', (done) => {
```
Comment (dignifiedquire) [inline src/pin.js]: This shouldn't be a test, it should be part of the `before` block

Comment (dignifiedquire) [inline src/pin.js]: Otherwise if you use `.only` on a test it is not run, resulting in errors.

File: src/pin.js [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+
+        expect(res).to.have.length(1)
+        expect(res[0].hash).to.equal(expectedMultihash)
+        expect(res[0].path).to.equal(expectedMultihash)
+        done()
+      })
+    })
+
+    describe('callback API', () => {
+      it('.rm (1st, because ipfs.files.add pins automatically)', (done) => {
```
Comment (dignifiedquire) [inline src/pin.js]: the `1st...` should be a comment rather than part of the test description

File: src/pin.js [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+          expect(err).to.not.exist
+          expect(res.Pins[0]).to.be.equal(hash)
+          done()
+        })
+      })
+
+      it('.list', (done) => {
+        ipfs.pin.ls((err, res) => {
+          expect(err).to.not.exist
+          expect(res).to.exist
```
Comment (dignifiedquire) [inline src/pin.js]: Needs more detailed testing, this could be anything in the result

File: src/pin.js [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+          })
+        })
+      })
+
+      it('.add', (done) => {
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.pin.add(hash, { recursive: false }, (err, res) => {
+          expect(err).to.not.exist
+          expect(res.Pins[0]).to.be.equal(hash)
```
Comment (dignifiedquire) [inline src/pin.js]: Can we place lower case these values? , i.e. `.pins`

File: src/pin.js [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+      it('.rm (1st, because ipfs.files.add pins automatically)', (done) => {
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.pin.rm(hash, { recursive: true }, (err, res) => {
+          expect(err).to.not.exist
+          expect(res).to.exist
+          ipfs.pin.ls({ type: 'direct' }, (err, res) => {
+            expect(err).to.not.exist
+            expect(res).to.exist
+            expect(res.Keys).to.be.empty
```
Comment (dignifiedquire) [inline src/pin.js]: please lower case `Keys` to `keys`

File: src/pin.js [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+            done()
+          })
+        })
+      })
+
+      it('.add', (done) => {
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.pin.add(hash, { recursive: false }, (err, res) => {
```
Comment (dignifiedquire) [inline src/pin.js]: what about `recursive: true`?

File: src/pin.js [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+          done()
+        })
+      })
+
+      it('.list for a specific hash', (done) => {
+        const hash = 'Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP'
+
+        ipfs.pin.ls(hash, (err, res) => {
+          expect(err).to.not.exist
+          expect(res).to.exist
```
Comment (dignifiedquire) [inline src/pin.js]: more detailed testing, that it actually returns the requested hash

File: src/pin.js (hunk 22 above)
Comment (diasdavid) [inline src/pin.js]: Good point!

**Comment by diasdavid** [comment]
Applied all the CR :)

File: API/pin/README.md [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+> Remove a hash from the pinset
+
+##### `Go` **WIP**
+
+##### `JavaScript` - ipfs.pin.rm(hash, [options, callback])
+
+Where:
+- `hash` is a multihash.
+- `options` is an object that can contain the following keys
+  - `recursive` - Recursively unpin the object linked. Type: bool. Default: 'false'
```
Comment (dignifiedquire) [inline API/pin/README.md]: 'false' -> `false`

File: package.json [hunk 30]
```diff
@@ -2,7 +2,7 @@
   "name": "interface-ipfs-core",
   "version": "0.8.0",
   "description": "A test suite and interface you can use to implement a IPFS core interface.",
-  "main": "lib/index.js",
+  "main": "src/index.js",
```
Comment (dignifiedquire) [inline package.json]: needs removal before merge

File: src/pin.js [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+        })
+      })
+
+      it('.ls type recursive', (done) => {
+        ipfs.pin.ls({ type: 'recursive' }, (err, pinset) => {
+          expect(err).to.not.exist
+          expect(Object.keys(pinset).length > 0).to.equal(true)
+          done()
+        })
+      })
```
Comment (dignifiedquire) [inline src/pin.js]: These three tests above could be simplified to

function testLs (type) {
  return it(`.ls type ${type}`, (done) => {
  })
}

File: API/pin/README.md (hunk 29 above)
Comment (diasdavid) [inline API/pin/README.md]: Richard disagreed on that one, I'll let that 🚲🏠 happen on a separate PR

File: src/pin.js (hunk 31 above)
Comment (diasdavid) [inline src/pin.js]: Simplified, but not simpler for changes, reading, etc :)

File: API/pin/README.md (hunk 29 above)
Comment (dignifiedquire) [inline API/pin/README.md]: @RichardLitt why? (can't find the comment), 'false' in a markdown document looks like it is a string.

File: src/pin.js (hunk 31 above)
Comment (dignifiedquire) [inline src/pin.js]: fair enough, but please use `.not.empty` rather than `.length > 0).equal(true)`

Comment (diasdavid) [inline src/pin.js]: you got it :)

**Comment by dignifiedquire** [comment]
lgtm

File: API/pin/README.md (hunk 29 above)
Comment (RichardLitt) [inline API/pin/README.md]: Yeah, that looks fine to me. This should be default true, though!
